FantasyUniverse/Services/Auth/FirebaseAuthService.swift

```diff
@@ -0,0 +1,144 @@
+                    
+                    self.currentUser = User(
+                        uid: user.uid,
+                        displayName: user.displayName,
+                        email: user.email,
+                        refreshToken: user.refreshToken)
+                    self.state = .success(with: self.currentUser!)
+                } else {
+                    self.currentUser = nil
+                    self.state = .unauthenticated
+                }
+            }
+    }
+    
+    func loginPublisher(with credentials: ManualSignInCredentials) -> AnyPublisher<Void, Error> {
+        Deferred {
+            Future { promise in
+                Auth
+                    .auth()
+                    .signIn(withEmail: credentials.email,
+                            password: credentials.password) { _, err in
+                        if let error = err {
+                            self.state = .failed(with: error)
+                            promise(.failure(error))
+                        } else {
+                            promise(.success(()))
+                        }
+                    }
+            }
+        }
+        .receive(on: RunLoop.main)
+        .eraseToAnyPublisher()
+    }
+    
+    func registrationPublisher(with credentials: SignUpCredentials) -> AnyPublisher<Void, Error> {
+        Deferred {
+            Future { promise in
+                Auth
+                    .auth()
+                    .createUser(withEmail: credentials.email,
+                                password: credentials.password) { res, err in
+                        if let error = err {
+                            promise(.failure(error))
+                        } else {
+                            if let uid = res?.user.uid {
+                                let values = [
+                                    AuthFirebaseKeys.firstname.rawValue:
+                                        credentials.firstname,
+                                    AuthFirebaseKeys.lastname.rawValue: credentials.lastname] as [String: Any]
+                                
+                                Database
+                                    .database()
+                                    .reference()
+                                    .child(AuthFirebaseKeys.users.rawValue)
+                                    .child(uid)
+                                    .updateChildValues(values) { error, _ in
+                                        if let err = error {
+                                            promise(.failure(err))
+                                        } else {
+                                            promise(.success(()))
+                                        }
+                                    }
+                                
+                            } else {
+                                promise(.failure(NSError(domain: "Invalid User Id", code: 0, userInfo: nil)))
```

Comment:
Make and enum like "FirebaseAuthError" with cases for errors like this `case invalidUserID`.